### DISCUSSION TITLE
Player colliding with NPCs, but NPCs not with each other

### DISCUSSION THREAD

**Discussion by zewa666** [Annoraaq/grid-engine#259] [Q&A] [answered]
So I was wondering how, using collision groups, one could define that the player collides with npcs, but NPCs do not collide with each other. NPCs are added dynamically so the only way I could think of doing it now is:

* Player has collisionGroup `player`
* When first NPC appears it has also `player` 
  * additionally a new group `npc1`
  * and the player is updated via `setCollisionGroups` to get `npc1` as well
* For every other NPC do same as previous step but with `npcN+1`

which results in the player having the same collision groups with every individual NPC, but they only share the group with the player and not each other. It sounds a bit weird though having to create so many groups, so I was wondering whether there might be a better approach to this?

**Comment by Annoraaq** [comment, marked as answer]
That's an interesting question. I can't think of a different way of achieving it with the current features.

Maybe there is a way to better support this in GridEngine in the future. Do you have any ideas?

**Reply by zewa666** [reply]
I could only think of a way to reduce the complexity from above by saying e.g. if the collisionGroups array of char includes `true` to make it intersect with any other group. So the player would have `true` and `"player"`.

**Reply by Annoraaq** [reply]
Do you mean the complexity in terms of asymptotic runtime complexity (Big O)?
To make this more efficient with a number of collision groups that scales with the amount of characters, the implementation could change by using HashSets instead of plain arrays: Take the smaller set and check for each element in`O(1)` whether it is member of the other set.

And the smaller set has size 1 in your case.

**Reply by zewa666** [reply]
I meant complexity in the perspective of the consumer (one less step to do), but certainly also runtime complexity. The later though is a question of how many NPCs one is going to have to be really noticable . also there might be different types of NPCs where A does collide with B (friendly vs enemy)

**Reply by zewa666** [reply]
another idea would be to define `collision` and `nonCollision` groups. say player char has `player` collision but `friend` nonCollision, now friendly NPCs could have `friend` nonCollision as well but enemies `player` as collision. nonCollision matches would negate collision effects. This way generation of new NPCs is would become trivial. the only open question is if both collision and nonCollision result in a match, which one wins :)

**Reply by Annoraaq** [reply]
A third idea would be to define which collision groups should collide with each other. Currently they collide with each other if they are the same. So `{"group1" : ["group1"]}`.

However, one could override this mapping. In the original example where NPCs should collide with the player but not with each other the following mapping would work:
 `{
   "player" : ["player", "npc"],
   "npc": ["player"]
 }`.
 
 This way "player"s would collide with "npc"s but "npc"s only with "player"s.
 
 GridEngine must assure that the relation stays [symmetric](https://en.wikipedia.org/wiki/Symmetric_relation).

**Reply by zewa666** [reply]
i think that would definitely work

**Reply by Annoraaq** [reply]
I created an issue for this so it does not get lost: #260
